exetera/processing/date_time_helpers.py

```diff
@@ -0,0 +1,172 @@
+    of days elapsed since this timestamp:
+    * If ``start_date`` is set, the start_date is used as the zero-date
+    * If ``start_date`` is not set:
+      * If ``date_filter`` is not set, the lowest timestamp is used as the zero-date
+      * If ``date_filter`` is set, the lowest unfiltered timestamp is used as the zero-date
+
+    As well as returning the elapsed days, this method can also return a filter for which
+    elapsed dates are valid. This is determined as follows:
+    * If ``date_filter``, ``start_date`` and ``end_date`` are None, None is returned
+    * otherwise:
+      * If ``date_filter`` is not provided, the filter represents all dates that are out
+        of range with respect to the start_date and end_date parameters
+      * If ``date_filter`` is provided, the filter is all dates out of range with respect to
+        the start_date and end_date parameters unioned with the date_filter that was
+        passed in
+
+    """
+    if start_date is None and end_date is None and date_filter is None:
+        min_date = date_field.min()
+        days = np.floor((date_field - min_date) / SECONDS_PER_DAY).astype(np.int32)
+        return days, None
+    else:
+        in_range = np.ones(len(date_field), dtype=bool) if date_filter is None else date_filter
+        if start_date is not None:
+            min_date = start_date
+            in_range = in_range & (date_field >= start_date)
+        else:
+            min_date = np.min(date_field if date_filter is None else date_field[date_filter])
+        if end_date is not None:
+            in_range = in_range & (date_field < end_date)
+        days = np.floor((date_field - min_date) / SECONDS_PER_DAY).astype(np.int32)
+        return days, in_range
+
+
+def generate_period_offset_map(periods: Sequence[datetime]
+                               ) -> ArrayLike[np.int32]:
+    """
+    Given a list of ordered datetimes relating to period boundaries, generate a numpy
+    array of days that map each day to a period.
+
+    Example:
+
+    .. code-block:: python
+
+      [datetime(2020,1,5), datetime(2020,1,12), datatime(2020,1,19), datetime(2020,1,26)]
+
+
+    generates the following output
+
+    .. code-block:: python
+
+      [0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2]
+
+    In the above example, each period spans a week, and periods cover a total of 3 weeks.
+    As a result, the output is 21 entries long, one for each day covered by the period, and
+    matches each day to the corresponding period.
+    """
+    period_deltas = [(p - periods[0]).days for p in periods]
+    period_per_day = np.zeros(period_deltas[-1], dtype=np.int32)
+    for i_p in range(len(period_deltas)-1):
+        period_per_day[period_deltas[i_p]:period_deltas[i_p + 1]] = i_p
+    return period_per_day
+
+
+def get_period_offsets(periods_by_day: ArrayLike[np.int32],
+                       days: ArrayLike[np.int32]
```

Comment:
`ArrayLike[np.int32]` occurs a lot, we should sometime put pre-defined types somewhere to reduce this duplication.

Reply:
We can't do generic types with ArrayLike, it appears